src/Core/Geometry/IndexedGeometry.hpp

```diff
@@ -0,0 +1,491 @@
+///
+///
+/// ## Accessing layers
+/// Each layer is also associated with a `lock` state, used to give read-only or to lock write
+/// access.
+///
+/// Layers can be accessed in different ways (see #containsLayer, #countLayers,
+/// #getFirstLayerOccurrence, and #getLayer):
+///  - query by name and semantics, by passing either #LayerKeyType or a pair of name/semantics
+///  - query by semantics (Utils::ObjectWithSemantic::SemanticNameCollection), names are ignored.
+///    Only the first occurrence found is returned when required.
+///  - query by semantic name (Utils::ObjectWithSemantic::SemanticName): matches any layer including
+///  the
+///    given semantic name. Only the first occurrence found is returned when required.
+///
+/// \note Layer ordering is arbitrary and might change each time a new layer is added.
+///
+///
+class RA_CORE_API MultiIndexedGeometry : public AttribArrayGeometry, public Utils::ObservableVoid
+{
+  public:
+    using LayerSemanticCollection = Utils::ObjectWithSemantic::SemanticNameCollection;
+    using LayerSemantic           = Utils::ObjectWithSemantic::SemanticName;
+    using LayerKeyType            = std::pair<LayerSemanticCollection, std::string>;
+
+    using PointAttribHandle  = AttribArrayGeometry::PointAttribHandle;
+    using NormalAttribHandle = AttribArrayGeometry::NormalAttribHandle;
+    using FloatAttribHandle  = AttribArrayGeometry::FloatAttribHandle;
+    using Vec2AttribHandle   = AttribArrayGeometry::Vec2AttribHandle;
+    using Vec3AttribHandle   = AttribArrayGeometry::Vec3AttribHandle;
+    using Vec4AttribHandle   = AttribArrayGeometry::Vec4AttribHandle;
+
+#ifdef MULTI_INDEX_MIMIC_TRIANGLE_MESH
+    using IndexType          = Vector3ui;
+    using IndexContainerType = VectorArray<IndexType>;
+#endif
+
+    inline MultiIndexedGeometry() = default;
+    explicit MultiIndexedGeometry( const MultiIndexedGeometry& other );
+    explicit MultiIndexedGeometry( MultiIndexedGeometry&& other );
+    explicit MultiIndexedGeometry( const AttribArrayGeometry& other );
+    explicit MultiIndexedGeometry( AttribArrayGeometry&& other );
+    MultiIndexedGeometry& operator=( const MultiIndexedGeometry& other );
+    MultiIndexedGeometry& operator=( MultiIndexedGeometry&& other );
+
+    void clear() override;
+
+    /// \brief Copy only the geometry and the indices from \p other, but not the attributes.
+    /// \see AttribArrayGeometry::copyBaseGeometry
+    void copy( const MultiIndexedGeometry& other );
+
+    /// \brief Check that the MultiIndexedGeometry is well built, asserting when it is not.
+    /// \note Only compiles to something when in debug mode.
+    void checkConsistency() const;
+
+    /// Append another MultiIndexedGeometry to this one. Layers with same
+    /// name/semantics are concatenated, and other layers are ignored
+    /// \return true if all fields have been copied
+    bool append( const MultiIndexedGeometry& other );
+
+#ifdef MULTI_INDEX_MIMIC_TRIANGLE_MESH
```

Comment:
Could this be moved to TriangleMesh class instead ?

Reply:
The TriangleMesh class is supposed to be removed, right ? 
So it does not change much, but I can do it for cosmetics.

Reply:
Fixed in 677b061ed41074bfa4225c3aa99fd3964f88328d